File: 0001_Two_Sum/main.cpp
class Solution {
  public:
    // T = O(n), S = O(n)
    vector<int> twoSum(vector<int> &nums, int target) {
        unordered_map<int, int> hash;
        for (int i = 0; i < nums.size(); i++) {
            int n = target - nums[i];
            if (hash.count(n)) {
                return {hash[n], i};
            }
            hash[nums[i]] = i;
        }
        return {-1, -1};
    }
};

File: 0001_Two_Sum/golang/main.go
package solution

func twoSum(nums []int, target int) []int {
	if len(nums) == 0 {
		return nil
	}
	hash := make(map[int]int)
	for i1, v1 := range nums {
		v2 := target - nums[i1]
		i2, ok := hash[v2]
		if ok {
			return []int{i1, i2}
		}
		hash[v1] = i1
	}
	return nil
}
